zerver/lib/actions.py

```diff
@@ -906,6 +907,44 @@ def do_send_messages(messages):
     # intermingle sending zephyr messages with other messages.
     return already_sent_ids + [message['message'].id for message in messages]
 
+def add_reaction(user_profile, message, emoji):
+    # type: (UserProfile, Message, RealmEmoji) -> None
+    reaction = Reaction()
+    reaction.sender = user_profile
+    reaction.message = message
+    reaction.emoji = emoji
+    reaction.save()
+
+    sender = {'id': user_profile.id, 'email': user_profile.email}
+
+    event = {'type': 'reaction',
+             'sender': sender,
+             'message_id': message.id,
+             'emoji_name': emoji.name} # type: Dict[str, Any]
+
+    ums = UserMessage.objects.filter(message=message.id)
+    send_event(event, [um.user_profile_id for um in ums])
+
+# check_add_reaction:
+# Checks the reaction and adds it
+def check_add_reaction(sender, emoji_name, message_id):
+    # type: (UserProfile, text_type, int) -> None
+    try:
+        message = Message.objects.select_related().get(id=message_id)
+    except Message.DoesNotExist:
+        raise JsonableError(_("Unknown message id"))
+    if emoji_name is None:
+        raise JsonableError(_("Missing emoji"))
+    if emoji_name == "":
+        raise JsonableError(_("Emoji name can't be empty"))
+    try:
+        # emoji = sender.realm.get_emoji()[emoji_name]
+        emoji = RealmEmoji.objects.get(realm=message.realm, name=emoji_name)
+    # except RealmEmoji.DoesNotExist:
+    except:
+        raise JsonableError(_("Emoji does not exist"))
```

Comment:
Check for non-realm emoji on disk; use `zerver.lib.bugdown.emoji_list` for this.  We will eventually want to clean up the data structure to be a set, not a list, but it's not super important.